File: apps/web/src/api/kyc.ts
import { apiJson } from "./client";

export type KycStatus = {
  verified: boolean;
  verified_at: string | null;
  required: boolean;
  provider: string | null;
};

export async function fetchKycStatus() {
  return apiJson<KycStatus>("/api/v1/kyc/status");
}

export type AdminKycUser = {
  user_id: string;
  phone: string;
  verified: boolean;
  verified_at: string | null;
  created_at: string;
};

export type KycAuditEntry = {
  audit_id: string;
  user_id: string;
  action: string;
  status: string;
  provider: string;
  message: string | null;
  real_name_masked: string | null;
  id_number_last4: string | null;
  created_at: string;
};

export async function fetchAdminKycPending() {
  return apiJson<AdminKycUser[]>("/api/v1/admin/kyc/pending");
}

export async function fetchAdminKycAudit(userId: string) {
  return apiJson<KycAuditEntry[]>(`/api/v1/admin/kyc/${userId}/audit`);
}

export async function submitKyc(realName: string, idNumber: string) {
  return apiJson<{ verified: boolean; message: string; audit_id: string }>("/api/v1/kyc/submit", {
    method: "POST",
    body: JSON.stringify({ real_name: realName, id_number: idNumber }),
  });
}

export async function adminVerifyKyc(userId: string, note?: string) {
  return apiJson<KycStatus>(`/api/v1/admin/kyc/${userId}/verify`, {
    method: "POST",
    body: JSON.stringify({ note }),
  });
}

export async function adminRevokeKyc(userId: string, note?: string) {
  return apiJson<KycStatus>(`/api/v1/admin/kyc/${userId}/revoke`, {
    method: "POST",
    body: JSON.stringify({ note }),
  });
}

File: apps/api/routes/kyc.py
from __future__ import annotations

from uuid import UUID

from apps.api.deps import get_current_user_id, get_session, require_admin_user
from apps.api.exceptions import raise_domain_http
from domains.kyc.service import KycService, KycServiceError
from fastapi import APIRouter, Depends, Request
from sqlalchemy.orm import Session
from voice_platform.kyc.schemas import (
    AdminKycUserSummary,
    AdminKycVerifyRequest,
    KycAuditEntry,
    KycStatusResponse,
    KycSubmitRequest,
    KycSubmitResponse,
)

router = APIRouter()


@router.get("/kyc/status", response_model=KycStatusResponse)
def kyc_status(
    user_id: UUID = Depends(get_current_user_id),
    session: Session = Depends(get_session),
) -> KycStatusResponse:
    try:
        return KycService(session).get_status(user_id)
    except KycServiceError as exc:
        raise_domain_http(exc)


@router.post("/kyc/submit", response_model=KycSubmitResponse)
def kyc_submit(
    body: KycSubmitRequest,
    user_id: UUID = Depends(get_current_user_id),
    session: Session = Depends(get_session),
) -> KycSubmitResponse:
    try:
        return KycService(session).submit(
            user_id,
            real_name=body.real_name,
            id_number=body.id_number,
        )
    except KycServiceError as exc:
        raise_domain_http(exc)


@router.get("/admin/kyc/pending", response_model=list[AdminKycUserSummary])
def admin_kyc_pending(
    limit: int = 50,
    _: UUID = Depends(require_admin_user),
    session: Session = Depends(get_session),
) -> list[AdminKycUserSummary]:
    return KycService(session).list_pending_users(limit=min(limit, 200))


@router.get("/admin/kyc/{user_id}/audit", response_model=list[KycAuditEntry])
def admin_kyc_audit(
    user_id: UUID,
    _: UUID = Depends(require_admin_user),
    session: Session = Depends(get_session),
) -> list[KycAuditEntry]:
    return KycService(session).list_audit(user_id)


@router.post("/admin/kyc/{user_id}/verify", response_model=KycStatusResponse)
def admin_kyc_verify(
    user_id: UUID,
    body: AdminKycVerifyRequest | None = None,
    _: UUID = Depends(require_admin_user),
    session: Session = Depends(get_session),
) -> KycStatusResponse:
    try:
        return KycService(session).admin_verify(user_id, note=body.note if body else None)
    except KycServiceError as exc:
        raise_domain_http(exc)


@router.post("/admin/kyc/{user_id}/revoke", response_model=KycStatusResponse)
def admin_kyc_revoke(
    user_id: UUID,
    body: AdminKycVerifyRequest | None = None,
    _: UUID = Depends(require_admin_user),
    session: Session = Depends(get_session),
) -> KycStatusResponse:
    try:
        return KycService(session).admin_revoke(user_id, note=body.note if body else None)
    except KycServiceError as exc:
        raise_domain_http(exc)


@router.post("/kyc/webhooks/saas", response_model=KycStatusResponse)
async def kyc_saas_webhook(
    request: Request,
    session: Session = Depends(get_session),
) -> KycStatusResponse:
    body = await request.body()
    signature = request.headers.get("X-Kyc-Webhook-Signature")
    try:
        return KycService(session).process_saas_webhook(body=body, signature=signature)
    except KycServiceError as exc:
        raise_domain_http(exc)
